utils/common/src/main/kotlin/EnvironmentVariableFilter.kt

```diff
@@ -0,0 +1,117 @@
+ *   some default patterns to match variables containing strings like "PASS", "USER", "TOKEN", etc.
+ * - There is an allow list to include variables even if they are matched by a pattern.
+ *
+ * So in order to determine whether a specific variable "E" can be passed to a child process, this filter applies the
+ * following steps:
+ * - If E is contained in the allow list, it is included.
+ * - Otherwise, E is included if and only if it is not matched by one of the exclusion patterns.
+ */
+object EnvironmentVariableFilter {
+    /**
+     * A set with patterns for variable names that are denied by default. All variables containing one of these
+     * strings (ignoring case) are not propagated to child processes.
+     */
+    val DEFAULT_DENY_PATTERNS = setOf(
+        "key",
+        "pass",
+        "pwd",
+        "token",
+        "user"
+    )
+
+    /** A set with known variable names that are accepted despite they are matched by deny patterns. */
+    val DEFAULT_ALLOW_NAMES = setOf(
+        "CARGO_HTTP_USER_AGENT",
+        "COMPOSER_ALLOW_SUPERUSER",
+        "CONAN_LOGIN_ENCRYPTION_KEY",
+        "CONAN_LOGIN_USERNAME",
+        "CONAN_PASSWORD",
+        "CONAN_USERNAME",
+        "CONAN_USER_HOME",
+        "CONAN_USER_HOME_SHORT",
+        "DOTNET_CLI_CONTEXT_ANSI_PASS_THRU",
+        "GIT_ASKPASS",
+        "GIT_HTTP_USER_AGENT",
+        "GRADLE_USER_HOME",
+        "HACKAGE_USERNAME",
+        "HACKAGE_PASSWORD",
+        "HACKAGE_KEY",
+        "PWD",
+        "USER",
+        "USERPROFILE"
+    )
+
+    /** Stores the current deny patterns used by this filter. */
+    private var denyPatterns = DEFAULT_DENY_PATTERNS.toPatterns()
+
+    /** Stores the current allow names used by this filter. */
+    private var allowNames = DEFAULT_ALLOW_NAMES
+
+    /**
+     * Reset this filter to use the given [denyPatterns] and [allowNames]. The [denyPatterns] are
+     * converted to regular expressions, so that they can be matched anywhere in variable names, ignoring case.
+     */
+    fun reset(
+        denyPatterns: Collection<String> = DEFAULT_DENY_PATTERNS,
+        allowNames: Collection<String> = DEFAULT_ALLOW_NAMES
+    ) {
+        this.denyPatterns = denyPatterns.toPatterns()
+        this.allowNames = allowNames.toSet()
+    }
+
+    /**
+     * Test whether the variable with the given [name] can be passed to a child process according to the criteria
+     * described in the header comment.
+     */
+    fun isAllowed(name: String): Boolean {
+        return name in allowNames || denyPatterns.none { it.matches(name) }
+    }
+
+    /**
+     * Remove all keys from [environment] that do not pass this filter.
+     */
+    fun filter(environment: MutableMap<String, String>): MutableMap<String, String> {
+        val keysToRemove = environment.keys.filterNot(EnvironmentVariableFilter::isAllowed)
+
+        environment -= keysToRemove.toSet()
+
+        return environment
```

Comment:
Shorter: `= environment.filter { (name, _) -> isAllowed(name) }`

Reply:
I find my original construct ugly, too. But the problem is that the filtering must be done in place instead of returning a new filtered map. This is because of the API of `ProcessBuilder` which returns a mutable map with environment variables, but does not allow setting this map.

Reply:
> But the problem is that the filtering must be done in place instead of returning a new filtered map.

Ah, right, sorry for missing that. Not sure if it's nicer, but this should work alternatively:

```
        val i = environment.iterator()
        while (i.hasNext()) {
            if (!isAllowed(i.next().key)) i.remove()
        }

        return environment
```

Actually, I guess I like your solution better as it's more compact. The only thing that bothered me is the `toSet()` conversion. I know that without it we get an inspection hint, but the hint is bogus in this case as there can be no duplicates in the list. So I'd indeed prefer your solution, but with the `toSet()` removed.